### PR TITLE
ARQ-1684: Fixes JaCoCo Extension Filters

### PR DESCRIPTION
**Pull request by hemanik**
#### Short description of what this resolves:

Fixes JaCoCo extension Filter for Includes/Excludes.

#### Changes proposed in this pull request:

- Moves filter composition from configuration class to its own component
- Includes unit tests for this new component applying single `ArchivePath` (test boundary conditions)
- Includes higher-level unit test where we apply filter on the ShrinkWrap archives (to cover broader scope of filtering on the collections)
- Fixes AND Filter as per the expected behaviour.

**Fixes**: #30

### CODE REVIEW

File: src/main/java/org/jboss/arquillian/extension/jacoco/client/filter/AndFilter.java [hunk 1]
```diff
@@ -16,13 +16,18 @@
     @Override
     public boolean include(T object)
     {
+        boolean include = false;
         for (Filter<T> f : filters)
         {
             if (!f.include(object))
             {
+                if (f.toString().contains("IncludeRegExpPaths") && include) {
```
Comment (hemanik) [inline src/main/java/org/jboss/arquillian/extension/jacoco/client/filter/AndFilter.java]: If there are multiple include patterns, we should include the asset matching any of them. 
The problem was that if the asset does not match other include patterns, it returns false, thus excluding the asset.

The check ensures that, if there are multiple include patterns and if a asset matches either of the include filter pattern, it should be included.

Comment (bartoszmajsak) [inline src/main/java/org/jboss/arquillian/extension/jacoco/client/filter/AndFilter.java]: Are any of the tests exposing this issue?

Comment (hemanik) [inline src/main/java/org/jboss/arquillian/extension/jacoco/client/filter/AndFilter.java]: Yes, [this test](https://github.com/arquillian/arquillian-extension-jacoco/pull/31/commits/b3a89f775131bb67e4f2a157b45555607dc59183#diff-3000a9437ea469d92e6b9063dc369723L127).

Comment (bartoszmajsak) [inline src/main/java/org/jboss/arquillian/extension/jacoco/client/filter/AndFilter.java]: I don't think this is ensuring correctness of this scenario. We probably have to have following structure instead:

I - set of include patterns
E - set of exclude patterns
```
E = e1 || e2 || .... || eN
I = i1 || i2 || ... || iN
```
the asset is included if  `!E && I`

Thus I think we should not only rely on `AndFilter` but improve the logic based on what this test exposed.

In addition, the generic filter (`AndFilter` in this case) shouldn't assume anything about the underlying implementations it's delegating to.

File: src/test/java/org/jboss/arquillian/extension/jacoco/test/unit/FilterComposerTestCase.java [hunk 2]
```diff
@@ -68,36 +68,36 @@ public void should_exclude_asset_not_matching_include_pattern() throws Exception
...
-   public void should_include_asset_not_matching_exclude_pattern() throws Exception
+   public void should_include_asset_matching_one_of_include_patterns() throws Exception
    {
       // given
-      final Filter<ArchivePath> filter = createComposedFilter(null, "org.arquillian.extension.jacoco.client.*");
+      final Filter<ArchivePath> filter = createComposedFilter("org.arquillian.extension.jacoco.client.*, org.arquillian.extension.integration.*", null);
 
       // when
-      final boolean include = filter.include(ArchivePaths.create("org/arquillian/extension/jacoco/container/CoverageDataCommand.class"));
+      final boolean include = filter.include(ArchivePaths.create("org/arquillian/extension/jacoco/client/ManifestAsset.class"));
```
Comment (bartoszmajsak) [inline src/test/java/org/jboss/arquillian/extension/jacoco/test/unit/FilterComposerTestCase.java]: This class is in different package now.

Comment (hemanik) [inline src/test/java/org/jboss/arquillian/extension/jacoco/test/unit/FilterComposerTestCase.java]: No, the class is still in the same package. I re-ordered a few test cases and now the diff shows something weird.

But what do you mean by refactoring friendly ?

Comment (bartoszmajsak) [inline src/test/java/org/jboss/arquillian/extension/jacoco/test/unit/FilterComposerTestCase.java]: https://github.com/arquillian/arquillian-extension-jacoco/blob/760571097351832b8a5a1d3cd9d36ea6a3fc9de7/src/main/java/org/jboss/arquillian/extension/jacoco/client/instrumentation/ManifestAsset.java is it?

Comment (bartoszmajsak) [inline src/test/java/org/jboss/arquillian/extension/jacoco/test/unit/FilterComposerTestCase.java]: By refactoring friendly I mean that we changed package of this class - which is some kind of refactoring. Now, since we rely on string representation for the path it does not reflect reality anymore as string path is wrong. We did that already for the higher level test. Please have a look at [adjusted test](https://github.com/arquillian/arquillian-extension-jacoco/pull/31/commits/fb488f328d7099c82a47b66ec51542fbd9356f58#diff-3000a9437ea469d92e6b9063dc369723R41) and make similar changes in all the other tests.
